### PR TITLE
crush: fix 'ceph osd crush reweight-subtree'

### PR DESCRIPTION
**Pull request by liewegas**
http://tracker.ceph.com/issues/11855

### CODE REVIEW

**Comment by dachary** [comment]
At the beginning of **CrushWrapper::adjust_subtree_weight** 
<pre>
   ldout(cct, 5) << "adjust_item_weight "
</pre>
could be fixed to include the method name instead while you're at it.

File: src/test/crush/CrushWrapper.cc [hunk 1]
```diff
@@ -470,6 +479,96 @@ TEST(CrushWrapper, adjust_item_weight) {
...
+  //c->dump_tree(&cout, NULL);
+
+  int r = c->adjust_subtree_weightf(g_ceph_context, host0, 2.0);
+  ASSERT_EQ(r, 2); // 2 items changed
+
+  //cout << "--------after---------" << std::endl;
+  //c->dump_tree(&cout, NULL);
+
+  ASSERT_EQ(c->get_bucket_weight(host0), 262144);
+  ASSERT_EQ(c->get_item_weight(host0), 262144);
```
Comment (dachary) [inline src/test/crush/CrushWrapper.cc]: Also 
<pre>
ASSERT_EQ(c->get_item_weight(rootno), XXX);
</pre>
to verify the weight changed up to the root ?

File: src/test/crush/CrushWrapper.cc [hunk 2]
```diff
@@ -470,6 +479,96 @@ TEST(CrushWrapper, adjust_item_weight) {
...
+    EXPECT_EQ(host_weight, c->get_bucket_weightf(bucket_id));
+
+    map<string,string> bloc;
+    bloc["root"] = "default";
+    EXPECT_EQ(0, c->insert_item(g_ceph_context, hostfake, host_weight,
+				FAKE, bloc));
+  }
+
+  //cout << "--------before---------" << std::endl;
+  //c->dump_tree(&cout, NULL);
```
Comment (dachary) [inline src/test/crush/CrushWrapper.cc]: Just to make the test easier to read, I would have liked the ASSERT_EQ below to be duplicated before the call as well, verifying the expected weight. It makes it trivial to understand the before / after transition.

**Comment by dachary** [comment]
Nitpicking, otherwise :+1: 

<pre>
Reviewed-by: Loic Dachary &lt;ldachary@redhat.com>
</pre>

**Comment by dachary** [comment]
@liewegas perfect :-)

**Comment by dragonylffly** [comment]
Regarding the algorithm, maybe it does not worth the effort to revise because it just runs once per user's command. For the returned 'changed', should it reflect the weight change of bucket, others looks perfect to me :+1:

**Comment by dachary** [comment]
@dragonylffly maybe it will be slow if someone has a very very large crushmap :-) But I think you're right, it's not an optimization that is required right now.

**Comment by liewegas** [comment]
On Wed, 10 Jun 2015, Li Wang wrote:
> Regarding the algorithm, maybe it does not worth the effort to revise because it just runs once per user's command. For the returned 'changed',
> should it reflect the weight change of bucket, others looks perfect to me :+1:

it's defined as the number of items changed currently.

thanks for the review!

**Pull request merged**